### PR TITLE
tt: add tt run module

### LINKED ISSUE #29: Implement `tt run`
`tt run` should to have the same functionality as `tt start`, except that it does not use `watchdog` and output should not be redirected, just write to `stdout`, `stderr`. The main purpose of this command is to check during development how the application will work under `tt`.
Also, in this mode. no one command of `tt` which work through `watchdog` can be used to interact with an instance.

### PR DESCRIPTION
**Pull request by better0fdead**
Added module to start instances without watchdog.

Closes #29

### CODE REVIEW

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank you for the patch.
- Add tests
- The `tt run` must be able to run a file (for example, `tt run example.lua`).
- `tt run` must be used to start the instance in `tt start`.
- I think the command should consider the config according to some flag (`--use-cfg`). I think we can use this in `tt start`. In this case, `watchdog` will need to watch for changes in the fields: `restart_on_failure`, `log_dir`. It may also be necessary to pass information to `tt run` (how exactly should be thought about) whether to create a socket and PIDfile or no.

File: cli/cmd/run.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"github.com/apex/log"
+	"github.com/spf13/cobra"
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/modules"
+	"github.com/tarantool/tt/cli/running"
+)
+
+// NewStartCmd creates start command.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go]: Fixed

File: cli/cmd/run.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+func NewRunCmd() *cobra.Command {
+	var startCmd = &cobra.Command{
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go]: Fixed

File: cli/cmd/run.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+			if err != nil {
+				log.Fatalf(err.Error())
+			}
+		},
+	}
+
+	return startCmd
+}
+
+// internalStartModule is a default start module.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go]: Fixed

File: cli/cmd/run.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/modules"
+	"github.com/tarantool/tt/cli/running"
+)
+
+// NewStartCmd creates start command.
+func NewRunCmd() *cobra.Command {
+	var startCmd = &cobra.Command{
+		Use:   "run [APPLICATION_NAME]",
+		Short: "Start tarantool instance without watchdog",
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go]: Fixed

File: cli/running/instance.go [hunk 5]
```diff
@@ -119,6 +119,44 @@ func (inst *Instance) Start() error {
 	return nil
 }
 
+// Run runs tarantool (using script/no script)
+func (inst *Instance) Run() error {
+	if inst.appPath == "" {
+		inst.Cmd = exec.Command("tarantool")
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go]: Fixed

File: cli/running/instance.go [hunk 6]
```diff
@@ -119,6 +119,44 @@ func (inst *Instance) Start() error {
...
+	} else {
+		inst.Cmd = exec.Command(inst.tarantoolPath, "-")
+	}
+
+	inst.Cmd.Stdout = os.Stdout
+	inst.Cmd.Stderr = os.Stderr
+
+	inst.Cmd.Env = append(os.Environ(), "TT_CLI_INSTANCE="+inst.appPath)
+	// Imitate the "tarantoolctl".
+	inst.Cmd.Env = append(inst.Cmd.Env, "TARANTOOLCTL=true")
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go]: Fixed

File: CHANGELOG.md [hunk 7]
```diff
@@ -23,3 +23,4 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Module ``tt rocks``, LuaRocks package manager.
 - Module ``tt cat``, to print into stdout the contents of .snap/.xlog files.
 - Module ``tt play``, to play the contents of .snap/.xlog files to another Tarantool instance.
+- Module ``tt run``, to start tarantool instance without watchdog.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply CHANGELOG.md]: Fixed

File: README.rst [hunk 8]
```diff
@@ -176,3 +176,4 @@ Common description. For a detailed description, use ``tt help command`` .
 * ``rocks`` - LuaRocks package manager.
 * ``cat`` - print into stdout the contents of .snap/.xlog files.
 * ``play`` - play the contents of .snap/.xlog files to another Tarantool instance.
+* ``run`` - start tarantool instance without watchdog.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply README.rst]: Fixed

File: cli/running/instance.go [hunk 9]
```diff
@@ -119,6 +119,44 @@ func (inst *Instance) Start() error {
 	return nil
 }
 
+// Run runs tarantool (using script/no script)
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go]: Fixed

File: cli/cmd/run.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			err := modules.RunCmd(&cmdCtx, cmd.Name(), &modulesInfo, internalRunModule, args)
+			if err != nil {
+				log.Fatalf(err.Error())
+			}
+		},
+	}
+
+	return startCmd
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go]: Added flag supports

File: cli/cmd/run.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// internalStartModule is a default start module.
+func internalRunModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
+
+	if len(args) != 0 {
+		if err := running.FillCtx(cliOpts, cmdCtx, args); err != nil {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go]: rewrote

File: cli/running/lua/launcher.lua [hunk 12]
```diff
@@ -126,7 +126,12 @@ local function start_instance()
         end
 
     end
-
+    if os.getenv("DUP2_STATE") then
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/lua/launcher.lua:129]: Added

File: cli/running/instance.go [hunk 13]
```diff
@@ -119,6 +120,91 @@ func (inst *Instance) Start() error {
...
+
+	inst.Cmd.Stdout = os.Stdout
+	inst.Cmd.Stderr = os.Stderr
+
+	inst.Cmd.Env = append(os.Environ(), "TT_CLI_INSTANCE="+inst.appPath)
+	// Set the sign that the program is running under "tt".
+	inst.Cmd.Env = append(inst.Cmd.Env, "TT_CLI=true")
+
+	if pipeFlag {
+		inst.Cmd.Env = append(inst.Cmd.Env, "DUP2_STATE=true")
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go:177]: Fixed

File: cli/cmd/run.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+		RunV: flagV,
+	}
+	return &cmdcontext.RunOpts{CmdCtx: &cmdCtx, RunFlags: &runFlags}
+}
+
+// NewRunCmd creates run command.
+func NewRunCmd() *cobra.Command {
+	var runCmd = &cobra.Command{
+		Use:   "run [APPLICATION_NAME]",
+		Short: "Start tarantool instance",
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go:41]: Fixed

File: cli/cmd/run.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+	return runCmd
+}
+
+// internalRunModule is a default run module.
+func internalRunModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
+	runOpts := getRunOpts(*cmdCtx)
+	if len(args) > 0 {
+		if args[0] == "-" {
+			if !terminal.IsTerminal(syscall.Stdin) {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go:62]: Added

File: cli/running/instance.go [hunk 16]
```diff
@@ -119,6 +120,91 @@ func (inst *Instance) Start() error {
 	return nil
 }
 
+// Run runs tarantool (using script/no script).
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go:123]: Fixed

File: cli/running/running.go [hunk 17]
```diff
@@ -323,6 +348,14 @@ func Stop(cmdCtx *cmdcontext.CmdCtx) error {
 	return nil
 }
 
+// Run starts an Instance (without watchdog).
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/running.go:351]: Changed

File: cli/cmd/run.go [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+			err := modules.RunCmd(&cmdCtx, cmd.Name(), &modulesInfo, internalRunModule, args)
+			if err != nil {
+				log.Fatalf(err.Error())
+			}
+		},
+	}
+	runCmd.Flags().StringVarP(&flagE, "evaluate", "e", "", "execute string 'EXPR'")
+	runCmd.Flags().StringVarP(&flagL, "library", "l", "", "require library 'NAME'")
+	runCmd.Flags().BoolVarP(&flagI, "interactive", "i", false, "enter interactive mode after executing 'SCRIPT'")
+	runCmd.Flags().BoolVarP(&flagV, "version", "v", false, "print tarantool version and exit")
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go:52]: Changed

File: cli/running/instance.go [hunk 19]
```diff
@@ -119,6 +120,91 @@ func (inst *Instance) Start() error {
...
+			inst.Cmd = exec.Command(inst.tarantoolPath)
+			inst.Cmd.Args = append(inst.Cmd.Args, command...)
+			inst.Cmd.Args = append(inst.Cmd.Args, "-")
+		} else {
+			inst.Cmd = exec.Command(inst.tarantoolPath)
+			inst.Cmd.Args = append(inst.Cmd.Args, command...)
+			inst.Cmd.Args = append(inst.Cmd.Args, "-")
+			inst.Cmd.Args = append(inst.Cmd.Args, flags.RunS)
+		}
+		syscall.Dup2(0, 7)
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go:162]: Added

File: cli/running/instance.go [hunk 20]
```diff
@@ -119,6 +120,91 @@ func (inst *Instance) Start() error {
...
+		StdinPipe.Close()
+	} else {
+		inst.Cmd.Env = append(inst.Cmd.Env, "DUP2_STATE=false")
+	}
+
+	// Start an Instance.
+	if err := inst.Cmd.Start(); err != nil {
+		return err
+	}
+	//inst.Cmd.Stdin = os.Stdin
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go:196]: Removed

File: cli/running/instance.go [hunk 21]
```diff
@@ -119,6 +120,91 @@ func (inst *Instance) Start() error {
...
+	}
+
+	inst.Cmd.Stdout = os.Stdout
+	inst.Cmd.Stderr = os.Stderr
+
+	inst.Cmd.Env = append(os.Environ(), "TT_CLI_INSTANCE="+inst.appPath)
+	// Set the sign that the program is running under "tt".
+	inst.Cmd.Env = append(inst.Cmd.Env, "TT_CLI=true")
+
+	if pipeFlag {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go:176]: Fixed

File: cli/running/instance.go [hunk 22]
```diff
@@ -119,6 +120,91 @@ func (inst *Instance) Start() error {
...
+	}
+	if flags.RunL != "" {
+		command = append(command, "-l")
+		command = append(command, flags.RunL)
+	}
+	if flags.RunV {
+		command = append(command, "-v")
+	}
+
+	pipeFlag := false
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go:141]: Added

File: cli/cmd/run.go [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/apex/log"
+	"github.com/spf13/cobra"
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/modules"
+	"github.com/tarantool/tt/cli/running"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go:17]: Added

File: cli/cmd/run.go [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+	flagE string
+	flagL string
+	flagI bool
+	flagS string
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go:21]: Renamed

File: cli/cmd/run.go [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+	flagV bool
+)
+
+func getRunOpts(cmdCtx_t cmdcontext.CmdCtx) *cmdcontext.RunOpts {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmd/run.go:25]: Fixed

File: cli/cmdcontext/cmdcontext.go [hunk 26]
```diff
@@ -67,3 +67,23 @@ type ConnectCtx struct {
...
+	// RunI contains "-i" flag content.
+	RunI bool
+	// RunS contains "-" flag content.
+	RunS string
+	//RunV contains "-v" flag content.
+	RunV bool
+}
+
+// RunOpts contains information for tt run.
+type RunOpts struct {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/cmdcontext/cmdcontext.go]: Moved

File: cli/running/instance.go [hunk 27]
```diff
@@ -119,6 +120,91 @@ func (inst *Instance) Start() error {
 	return nil
 }
 
+// Run runs tarantool (using script/no script).
+func (inst *Instance) Run(flags *cmdcontext.RunFlags) error {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go]: Refactored

File: cli/running/run_test.go [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+	"os/exec"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/tarantool/tt/cli/ttlog"
+)
+
+// startTestInstance starts instance for the test.
```
Comment (LeonidVas) [inline cli/running/run_test.go:16]: `startTestInstance` -> `startTestRun`

File: cli/running/run_test.go [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+func startTestRun(t *testing.T, app string, consoleSock string, logger *ttlog.Logger, flags *RunFlags) *Instance {
+
+	assert := assert.New(t)
+	appPath := ""
+	if app != "" {
+		appPath := path.Join(instTestAppDir, app+".lua")
+		_, err := os.Stat(appPath)
+		//fmt.Printf(appPath)
```
Comment (LeonidVas) [inline cli/running/run_test.go:24]: Clean the code.

File: cli/running/run_test.go [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+// Instance that remain alive after the test done.
+func cleanupTestRun(inst *Instance) {
+	if inst.IsAlive() {
+		inst.Cmd.Process.Kill()
+	}
+	if _, err := os.Stat(inst.consoleSocket); err == nil {
+		os.Remove(inst.consoleSocket)
+	}
+}
+func TestEflag(t *testing.T) {
```
Comment (LeonidVas) [inline cli/running/run_test.go:51]: Use normal names instead of S,E...

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Points from https://github.com/tarantool/tt/pull/73#pullrequestreview-1026093609 are relevant.

Not working scenarios:
* Forwarding of arguments:
image

* Stop an instance with SIGINT
image

**Comment by better0fdead** [comment]
> Points from [#73 (review)](https://github.com/tarantool/tt/pull/73#pullrequestreview-1026093609) are relevant.
> 
> Not working scenarios:
> 
> * Forwarding of arguments:
>   image
> * Stop an instance with SIGINT
>   image

Fixed

File: test/integration/run/test_run.py [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+        stdout=subprocess.PIPE,
+        text=True
+    )
+    start_output = instance_process.stdout.readline()
+    assert re.search(r"   • Running an instance...   \n", start_output)
+    run_output = instance_process.stdout.readline()
+    assert re.search(r"123", run_output)
+
+
+def test_running_flag_v(tt_cmd, tmpdir):
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/integration/run/test_run.py]: Renamed

File: cli/running/run_test.go [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+	assert := assert.New(t)
+
+	flags := &RunFlags{RunEval: "print('123')", RunLib: "", RunInteractive: false, RunStdin: "", RunVersion: false}
+	old := os.Stdout
+	r, w, _ := os.Pipe()
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/run_test.go]: Changed

File: cli/running/run_test.go (hunk 30 above)
Comment (better0fdead) [reply cli/running/run_test.go]: Changed

File: cli/running/run_test.go [hunk 33]
```diff
@@ -0,0 +1,136 @@
...
+
+	assert.Nilf(err, `Can't create an instance. Error: "%v".`, err)
+
+	err = inst.Run(flags)
+	assert.Nilf(err, `Can't start the instance. Error: "%v".`, err)
+
+	return inst
+}
+
+// cleanupTestInstance sends a SIGKILL signal to test
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/run_test.go]: Changed

File: cli/running/running.go [hunk 34]
```diff
@@ -218,53 +238,61 @@ func createPIDFile(pidFileName string) error {
...
 		if len(args) > 1 {
 			return fmt.Errorf("Currently, you can specify only one instance at a time.")
 		} else {
 			return fmt.Errorf("Please specify the name of the application.")
 		}
 	}
-
 	appName := args[0]
+	if runFlag {
+		if len(appName) > 4 {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/running.go]: Changed

File: cli/running/running.go [hunk 35]
```diff
@@ -218,53 +238,61 @@ func createPIDFile(pidFileName string) error {
...
 		return fmt.Errorf("Can't find an application init file: %s", err)
 	}
 
 	cmdCtx.Running.AppPath = appPath
 
-	runDir := cliOpts.App.RunDir
-	if runDir == "" {
-		if runDir, err = os.Getwd(); err != nil {
-			return fmt.Errorf(`Can't get the "RunDir: %s"`, err)
+	if !runFlag {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/running.go]: Changed

File: cli/running/lua/launcher.lua [hunk 36]
```diff
@@ -127,6 +127,13 @@ local function start_instance()
 
     end
 
+    -- If stdin of the program was moved, then move it back.
+    if os.getenv("TT_DUP2_STATE") then
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/lua/launcher.lua]: Changed

File: cli/running/lua/launcher.lua [hunk 37]
```diff
@@ -127,6 +127,13 @@ local function start_instance()
 
     end
 
+    -- If stdin of the program was moved, then move it back.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/lua/launcher.lua]: Changed

File: cli/running/instance.go [hunk 38]
```diff
@@ -119,6 +119,91 @@ func (inst *Instance) Start() error {
...
+			return err
+		}
+		if inst.appPath == "" {
+			StdinPipe.Write([]byte(flags.RunStdin))
+		} else {
+			StdinPipe.Write([]byte(instanceLauncher))
+		}
+		StdinPipe.Close()
+	} else {
+		inst.Cmd.Env = append(inst.Cmd.Env, "TT_DUP2_STATE=false")
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go]: Changed

File: test/integration/run/test_app/test_app.lua [hunk 39]
```diff
@@ -0,0 +1,3 @@
+print(123)
+print(123)
+print(123)
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply test/integration/run/test_app/test_app.lua]: Changed

File: cli/running/run_test.go (hunk 28 above)
Comment (better0fdead) [reply cli/running/run_test.go]: Changed

File: cli/running/run_test.go (hunk 29 above)
Comment (better0fdead) [reply cli/running/run_test.go]: Cleaned

File: cli/running/running.go [hunk 40]
```diff
@@ -218,53 +238,61 @@ func createPIDFile(pidFileName string) error {
 }
 
 // FillCtx fills the RunningCtx context.
-func FillCtx(cliOpts *config.CliOpts, cmdCtx *cmdcontext.CmdCtx, args []string) error {
+func FillCtx(cliOpts *config.CliOpts, cmdCtx *cmdcontext.CmdCtx, args []string, runFlag bool) error {
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/running.go]: Done

File: cli/running/instance.go [hunk 41]
```diff
@@ -119,6 +119,91 @@ func (inst *Instance) Start() error {
...
+	if flags.RunVersion {
+		command = append(command, "-v")
+	}
+	return command
+}
+
+// Run runs tarantool instance.
+func (inst *Instance) Run(flags *RunFlags) error {
+	command := makeRunCommand(flags)
+	// pipeFlag is a flag used to indicate whether stdin should be moved or not.
```
[the remark this answers is not in the document]
Comment (better0fdead) [reply cli/running/instance.go]: Changed

**Comment by better0fdead** [comment]
[#1](https://github.com/tarantool/tt/pull/73#discussion_r919028513) Added comment
[#2](https://github.com/tarantool/tt/pull/73#discussion_r919029819) I can explain it in call, in shortly it is needed because in case when we need to find app file we need already run0pts.
[#3](https://github.com/tarantool/tt/pull/73#discussion_r919057585) Changed

File: cli/cmdcontext/cmdcontext.go [hunk 42]
```diff
@@ -3,9 +3,15 @@ package cmdcontext
...
 type CmdCtx struct {
-	Cli     CliCtx
+	// Cli - CLI context. Contains flags passed when starting
+	// Tarantool CLI and some other parameters.
+	Cli CliCtx
+	// Running contains information for running an application instance.
 	Running RunningCtx
+	// Connect contains information for connecting to the instance.
 	Connect ConnectCtx
+	// CommandName contains name of the command
```
Comment (LeonidVas) [inline cli/cmdcontext/cmdcontext.go:13]: Set a dot at the end of the sentence.

File: cli/running/running.go [hunk 43]
```diff
@@ -218,23 +240,32 @@ func createPIDFile(pidFileName string) error {
 }
 
 // FillCtx fills the RunningCtx context.
-func FillCtx(cliOpts *config.CliOpts, cmdCtx *cmdcontext.CmdCtx, args []string) error {
+func FillCtx(cliOpts *config.CliOpts, cmdCtx *cmdcontext.CmdCtx,
```
Comment (LeonidVas) [inline cli/running/running.go:243]: Unnecessary changes.

File: cli/running/run_test.go [hunk 44]
```diff
@@ -0,0 +1,146 @@
...
+		"The message in the log is different from what was expected.")
+}
+
+func TestStdinflag(t *testing.T) {
+	assert := assert.New(t)
+
+	flags := &RunFlags{RunEval: "", RunLib: "",
+		RunInteractive: false, RunStdin: "print('123')", RunVersion: false}
+	old := os.Stdout
+	r, w, _ := os.Pipe()
```
Comment (LeonidVas) [inline cli/running/run_test.go:76]: Do not use one-letter names,

File: cli/running/run_test.go [hunk 45]
```diff
@@ -0,0 +1,146 @@
...
+		"The message in the log is different from what was expected.")
+}
+
+func TestVersionflag(t *testing.T) {
+	assert := assert.New(t)
+
+	flags := &RunFlags{RunEval: "", RunLib: "",
+		RunInteractive: false, RunStdin: "", RunVersion: true}
+	old := os.Stdout
+	r, w, _ := os.Pipe()
```
Comment (LeonidVas) [inline cli/running/run_test.go:95]: Do not use one-letter names,

File: cli/running/run_test.go [hunk 46]
```diff
@@ -0,0 +1,146 @@
...
+		"The message in the log is different from what was expected.")
+}
+
+func TestInteractiveflag(t *testing.T) {
+	assert := assert.New(t)
+
+	flags := &RunFlags{RunEval: "", RunLib: "",
+		RunInteractive: true, RunStdin: "", RunVersion: false}
+	old := os.Stdout
+	r, w, _ := os.Pipe()
```
Comment (LeonidVas) [inline cli/running/run_test.go:114]: Do not use one-letter names,

File: cli/running/run_test.go [hunk 47]
```diff
@@ -0,0 +1,146 @@
...
+	inst.Stop(50 * time.Millisecond)
+}
+
+func TestLibflag(t *testing.T) {
+	assert := assert.New(t)
+
+	flags := &RunFlags{RunEval: "print('123')", RunLib: "os",
+		RunInteractive: false, RunStdin: "", RunVersion: false}
+	old := os.Stdout
+	r, w, _ := os.Pipe()
```
Comment (LeonidVas) [inline cli/running/run_test.go:134]: Do not use one-letter names,

File: test/integration/run/test_app/test_app.lua [hunk 48]
```diff
@@ -0,0 +1 @@
+print('Instance running!')
```
Comment (LeonidVas) [inline test/integration/run/test_app/test_app.lua:1]: image

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank you for the patchset.
The code already looks worthy.
The case when the code does not work correctly:
```
┌─[leonid@vasya-hp]─[~/work/mail/tt]
└──╼ cat test_2.lua 
print(arg[1])
┌─[leonid@vasya-hp]─[~/work/mail/tt]
└──╼ cat test_2.lua | tarantool - --my-app-option
--my-app-option
┌─[leonid@vasya-hp]─[~/work/mail/tt]
└──╼ cat test_2.lua | ./tt run - --my-app-option
unknown flag: --my-app-option

Usage: tt run [flags] 

```

**Comment by LeonidVas** [comment]
If any commit comes after the closing commit, it is `Follow ups #', not `Part of #'.
image

File: cli/cmdcontext/cmdcontext.go (hunk 42 above)
Comment (better0fdead) [reply cli/cmdcontext/cmdcontext.go]: Done

File: cli/running/running.go (hunk 43 above)
Comment (better0fdead) [reply cli/running/running.go:243]: It is for new linter.(line is too long)

File: cli/running/run_test.go (hunk 44 above)
Comment (better0fdead) [reply cli/running/run_test.go]: Done

File: cli/running/run_test.go (hunk 45 above)
Comment (better0fdead) [reply cli/running/run_test.go]: Done

File: cli/running/run_test.go (hunk 46 above)
Comment (better0fdead) [reply cli/running/run_test.go]: Done

File: cli/running/run_test.go (hunk 47 above)
Comment (better0fdead) [reply cli/running/run_test.go]: Done

**Comment by better0fdead** [comment]
> If any commit comes after the closing commit, it is `Follow ups #', not `Part of #'.
> image
Fixed

**Comment by better0fdead** [comment]
> Hi! Thank you for the patchset. The code already looks worthy. The case when the code does not work correctly:
> 
> ```
> ┌─[leonid@vasya-hp]─[~/work/mail/tt]
> └──╼ cat test_2.lua 
> print(arg[1])
> ┌─[leonid@vasya-hp]─[~/work/mail/tt]
> └──╼ cat test_2.lua | tarantool - --my-app-option
> --my-app-option
> ┌─[leonid@vasya-hp]─[~/work/mail/tt]
> └──╼ cat test_2.lua | ./tt run - --my-app-option
> unknown flag: --my-app-option
> 
> Usage: tt run [flags] 
> ```

Added comment about it in help of command.

**Review by LeonidVas — APPROVED** [review]
LGTM.

**Pull request merged**